README.md
[//]: # "This README.md file is auto-generated, all changes to this file will be lost."
[//]: # "To regenerate it, use `python -m synthtool`."
<img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>

# [Google Analytics Data: Node.js Client](https://github.com/googleapis/nodejs-analytics-data)

[![release level](https://img.shields.io/badge/release%20level-general%20availability%20%28GA%29-brightgreen.svg?style=flat)](https://cloud.google.com/terms/launch-stages)
[![npm version](https://img.shields.io/npm/v/@google-analytics/data.svg)](https://www.npmjs.org/package/@google-analytics/data)
[![codecov](https://img.shields.io/codecov/c/github/googleapis/nodejs-analytics-data/master.svg?style=flat)](https://codecov.io/gh/googleapis/nodejs-analytics-data)




Data client for Node.js


A comprehensive list of changes in each version may be found in
[the CHANGELOG](https://github.com/googleapis/nodejs-analytics-data/blob/master/CHANGELOG.md).

* [Google Analytics Data Node.js Client API Reference][client-docs]
* [Google Analytics Data Documentation][product-docs]
* [github.com/googleapis/nodejs-analytics-data](https://github.com/googleapis/nodejs-analytics-data)

Read more about the client libraries for Cloud APIs, including the older
Google APIs Client Libraries, in [Client Libraries Explained][explained].

[explained]: https://cloud.google.com/apis/docs/client-libraries-explained

**Table of contents:**


* [Quickstart](#quickstart)
  * [Before you begin](#before-you-begin)
  * [Installing the client library](#installing-the-client-library)
  * [Using the client library](#using-the-client-library)
* [Samples](#samples)
* [Versioning](#versioning)
* [Contributing](#contributing)
* [License](#license)

## Quickstart

### Before you begin

1.  [Select or create a Cloud Platform project][projects].
1.  [Enable the Google Analytics Data API][enable_api].
1.  [Set up authentication with a service account][auth] so you can access the
    API from your local workstation.

### Installing the client library

```bash
npm install @google-analytics/data
```


### Using the client library

```javascript
/**
 * TODO(developer): Uncomment this variable and replace with your GA4
 *   property ID before running the sample.
 */
// const propertyId = 'YOUR-GA4-PROPERTY-ID';

// Imports the Google Analytics Data API client library
const {AlphaAnalyticsDataClient} = require('@google-analytics/data');

// Creates a client
const client = new AlphaAnalyticsDataClient();

// Runs a simple report.
async function runReport() {
  const [response] = await client.runReport({
    entity: {
      propertyId: propertyId,
    },
    dateRanges: [
      {
        startDate: '2020-03-31',
        endDate: 'today',
      },
    ],
    dimensions: [
      {
        name: 'city',
      },
    ],
    metrics: [
      {
        name: 'activeUsers',
      },
    ],
  });

  console.log('Report result:');
  response.rows.forEach(row => {
    console.log(row.dimensionValues[0], row.metricValues[0]);
  });
}

runReport();

```



## Samples

Samples are in the [`samples/`](https://github.com/googleapis/nodejs-analytics-data/tree/master/samples) directory. The samples' `README.md`
has instructions for running the samples.

| Sample                      | Source Code                       | Try it |
| --------------------------- | --------------------------------- | ------ |
| Quickstart | [source code](https://github.com/googleapis/nodejs-analytics-data/blob/master/samples/quickstart.js) | [![Open in Cloud Shell][shell_img]](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/googleapis/nodejs-analytics-data&page=editor&open_in_editor=samples/quickstart.js,samples/README.md) |



The [Google Analytics Data Node.js Client API Reference][client-docs] documentation
also contains samples.

## Supported Node.js Versions

Our client libraries follow the [Node.js release schedule](https://nodejs.org/en/about/releases/).
Libraries are compatible with all current _active_ and _maintenance_ versions of
Node.js.

Client libraries targetting some end-of-life versions of Node.js are available, and
can be installed via npm [dist-tags](https://docs.npmjs.com/cli/dist-tag).
The dist-tags follow the naming convention `legacy-(version)`.

_Legacy Node.js versions are supported as a best effort:_

* Legacy versions will not be tested in continuous integration.
* Some security patches may not be able to be backported.
* Dependencies will not be kept up-to-date, and features will not be backported.

#### Legacy tags available

* `legacy-8`: install client libraries from this dist-tag for versions
  compatible with Node.js 8.

## Versioning

This library follows [Semantic Versioning](http://semver.org/).


This library is considered to be **General Availability (GA)**. This means it
is stable; the code surface will not change in backwards-incompatible ways
unless absolutely necessary (e.g. because of critical security issues) or with
an extensive deprecation period. Issues and requests against **GA** libraries
are addressed with the highest priority.





More Information: [Google Cloud Platform Launch Stages][launch_stages]

[launch_stages]: https://cloud.google.com/terms/launch-stages

## Contributing

Contributions welcome! See the [Contributing Guide](https://github.com/googleapis/nodejs-analytics-data/blob/master/CONTRIBUTING.md).

Please note that this `README.md`, the `samples/README.md`,
and a variety of configuration files in this repository (including `.nycrc` and `tsconfig.json`)
are generated from a central template. To edit one of these files, make an edit
to its template in this
[directory](https://github.com/googleapis/synthtool/tree/master/synthtool/gcp/templates/node_library).

## License

Apache Version 2.0

See [LICENSE](https://github.com/googleapis/nodejs-analytics-data/blob/master/LICENSE)

[client-docs]: https://googleapis.dev/nodejs/analytics-data/latest/index.html
[product-docs]: https://developers.google.com/analytics/trusted-testing/analytics-data
[shell_img]: https://gstatic.com/cloudssh/images/open-btn.png
[projects]: https://console.cloud.google.com/project
[billing]: https://support.google.com/cloud/answer/6293499#enable-billing
[enable_api]: https://console.cloud.google.com/flows/enableapi?apiid=analyticsdata.googleapis.com
[auth]: https://cloud.google.com/docs/authentication/getting-started


CHANGELOG.md
# Changelog

### [2.1.1](https://www.github.com/googleapis/nodejs-analytics-data/compare/v2.1.0...v2.1.1) (2021-02-12)


### Bug Fixes

* **deps:** update dependency google-auth-library to v7 ([#58](https://www.github.com/googleapis/nodejs-analytics-data/issues/58)) ([77002d8](https://www.github.com/googleapis/nodejs-analytics-data/commit/77002d877d4c353864d0a55f1247855e80e42fab))

## [2.1.0](https://www.github.com/googleapis/nodejs-analytics-data/compare/v2.0.0...v2.1.0) (2021-01-09)


### Features

* adds style enumeration ([#47](https://www.github.com/googleapis/nodejs-analytics-data/issues/47)) ([3028672](https://www.github.com/googleapis/nodejs-analytics-data/commit/3028672c3307fc68aea858c5d7f901f606a73a99))

## [2.0.0](https://www.github.com/googleapis/nodejs-analytics-data/compare/v1.1.0...v2.0.0) (2020-12-28)


### ⚠ BREAKING CHANGES

* GetUniversalMetadata method removed from the API, GetMetadata method should be used instead feat: reporting requests support longer date ranges (#43)

### Bug Fixes

* GetUniversalMetadata method removed from the API, GetMetadata method should be used instead feat: reporting requests support longer date ranges ([#43](https://www.github.com/googleapis/nodejs-analytics-data/issues/43)) ([661ee0e](https://www.github.com/googleapis/nodejs-analytics-data/commit/661ee0e27abea9978d09ee06b2d6c259712b351f))

## [1.1.0](https://www.github.com/googleapis/nodejs-analytics-data/compare/v1.0.2...v1.1.0) (2020-11-11)


### Features

* added RunRealtimeReport method that returns a customized report of realtime event data for a GA4 property docs: minor documentation updates ([#29](https://www.github.com/googleapis/nodejs-analytics-data/issues/29)) ([58da0fe](https://www.github.com/googleapis/nodejs-analytics-data/commit/58da0fe05418d91db1bbb698d787d3863266ff30))

### [1.0.2](https://www.github.com/googleapis/nodejs-analytics-data/compare/v1.0.1...v1.0.2) (2020-11-07)


### Bug Fixes

* do not modify options object, use defaultScopes ([#30](https://www.github.com/googleapis/nodejs-analytics-data/issues/30)) ([f5c9f64](https://www.github.com/googleapis/nodejs-analytics-data/commit/f5c9f64b3074a41ba590ce8be9aadd9961c0ab57))

### [1.0.1](https://www.github.com/googleapis/nodejs-analytics-data/compare/v1.0.0...v1.0.1) (2020-11-03)


### Bug Fixes

* **deps:** update dependency @google-cloud/billing to v2 ([#7](https://www.github.com/googleapis/nodejs-analytics-data/issues/7)) ([de83230](https://www.github.com/googleapis/nodejs-analytics-data/commit/de832308358059c3dbd59a4e9b118b5c2a208570))

## 1.0.0 (2020-09-02)


### ⚠ BREAKING CHANGES

* generate first stub of library (#1)

### Features

* generate first stub of library ([#1](https://www.github.com/googleapis/nodejs-analytics-data/issues/1)) ([5514d1c](https://www.github.com/googleapis/nodejs-analytics-data/commit/5514d1cc64de222e29ea6c3e6c3ca6f80bce9eea))


samples/quickstart_oauth2.js
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

/** This application demonstrates the usage of the Analytics Data API using
 OAuth2 credentials.
 Please familiarize yourself with the OAuth2 flow guide at
 https://developers.google.com/identity/protocols/oauth2

 For more information on authenticating as an end user, see
 https://cloud.google.com/docs/authentication/end-user

 In a nutshell, you need to:

 1. Create your OAuth2 client credentials in Google Cloud Console.
 Choose "Web application" when asked for an application type.

 https://support.google.com/cloud/answer/6158849

 2. When configuring the web application credentials, add
 "http://localhost:3000/oauth2callback" to "Authorized redirect URIs".
 In case a different callback URI is desired, this application needs to be
 updated accordingly.

 3. Download a credentials file using "Download JSON" button in the credentials
 configuration dialog and save it as `oauth2.keys.json` in the same
 directory with this sample app.

 4. Uncomment `propertyId` variable in main() function below and specify the
 value of the property id you want to access.

 5. Run the following commands from the current directory in order to install
 dependencies and run the sample app:

 npm install
 cd ..
 node samples/quickstart_oauth2.js
 */

function main(propertyId = 'YOUR-GA4-PROPERTY-ID') {
  // [START analytics_data_quickstart_oauth2]
  /**
   * TODO(developer): Uncomment this variable and replace with your GA4
   *   property ID before running the sample.
   */
  // propertyId = 'YOUR-GA4-PROPERTY-ID';

  //Imports the Google Analytics Data API client library.
  const {AlphaAnalyticsDataClient} = require('@google-analytics/data');

  const {OAuth2Client} = require('google-auth-library');
  const {grpc} = require('google-gax');
  const http = require('http');
  const url = require('url');
  const open = require('open');
  const destroyer = require('server-destroy');

  // Reads the secrets from a `oauth2.keys.json` file, downloaded from the
  // Google Developers Console.
  // eslint-disable-next-line node/no-unpublished-require
  // eslint-disable-next-line node/no-missing-require, node/no-unpublished-require
  const keys = require('./oauth2.keys.json');

  // This sample app only calls read-only methods from the Data API. Include
  // additional scopes if calling methods that modify the configuration.
  const SCOPES = ['https://www.googleapis.com/auth/analytics.readonly'];

  // Returns the Analytics Data API client set up using the provided OAuth2
  // credentials.
  function getAnalyticsDataClient(authClient) {
    const sslCreds = grpc.credentials.createSsl();
    const credentials = grpc.credentials.combineChannelCredentials(
      sslCreds,
      grpc.credentials.createFromGoogleCredential(authClient)
    );
    return new AlphaAnalyticsDataClient({
      sslCreds: credentials,
    });
  }

  /**
   * Creates a new OAuth2Client, and goes through the OAuth2 content
   * workflow.  Returns the client to the callback.
   */
  function getOAuth2Client() {
    return new Promise((resolve, reject) => {
      // Create an OAuth client to authorize the API call.
      const oAuth2Client = new OAuth2Client(
        keys.web.client_id,
        keys.web.client_secret,
        // A hardcoded callback URL used in this sample app.
        'http://localhost:3000/oauth2callback'
      );

      // Generate the url that will be used for the consent dialog.
      const authorizeUrl = oAuth2Client.generateAuthUrl({
        access_type: 'offline',
        scope: SCOPES.join(' '),
      });

      // Open an http server to accept the oauth callback. In this example, the
      // only request to our webserver is to /oauth2callback?code=<code>
      const server = http
        .createServer(async (req, res) => {
          try {
            if (req.url.indexOf('/oauth2callback') > -1) {
              // Acquire the code from the querystring, and close the web
              // server.
              const qs = new url.URL(req.url, 'http://localhost:3000')
                .searchParams;
              const code = qs.get('code');
              console.log(`Code is ${code}`);
              res.end(
                'Authentication successful! Please return to the console.'
              );
              server.destroy();

              // Now that we have the code, use that to acquire tokens.
              const r = await oAuth2Client.getToken(code);
              // Make sure to set the credentials on the OAuth2 client.
              oAuth2Client.setCredentials(r.tokens);
              console.info('Tokens acquired.');
              resolve(oAuth2Client);
            }
          } catch (e) {
            reject(e);
          }
        })
        .listen(3000, () => {
          // Open the browser to the authorize url to start the workflow.
          // This line will not work if you are running the code in the
          // environment where a browser is not available. In this case,
          // copy the URL and open it manually in a browser.
          console.info(`Opening the browser with URL: ${authorizeUrl}`);
          open(authorizeUrl, {wait: false}).then(cp => cp.unref());
        });
      destroyer(server);
    });
  }

  // Runs a simple report using the supplied Data API client instance.
  async function runReport() {
    // Starts the OAuth2 flow and runs the report.
    const oAuth2Client = await getOAuth2Client();
    const analyticsDataClient = getAnalyticsDataClient(oAuth2Client);

    const [response] = await analyticsDataClient.runReport({
      entity: {
        propertyId: propertyId,
      },
      dateRanges: [
        {
          startDate: '2020-03-31',
          endDate: 'today',
        },
      ],
      dimensions: [
        {
          name: 'city',
        },
      ],
      metrics: [
        {
          name: 'activeUsers',
        },
      ],
    });

    console.log('Report result:');
    response.rows.forEach(row => {
      console.log(row.dimensionValues[0], row.metricValues[0]);
    });
  }

  runReport();

  // [END analytics_data_quickstart_oauth2]
}

process.on('unhandledRejection', err => {
  console.error(err.message);
  process.exitCode = 1;
});
main(...process.argv.slice(2));
